overlord/ifacestate/handlers.go

```diff
@@ -1173,18 +1203,44 @@ func (m *InterfaceManager) doAutoConnect(task *state.Task, _ *tomb.Tomb) error {
 		}
 	}
 
-	autots, err := batchConnectTasks(st, snapsup, newconns, connOpts)
+	autots, hasHooks, err := batchConnectTasks(st, snapsup, newconns, connOpts)
 	if err != nil {
 		return err
 	}
 
-	if m.preseed && len(autots.Tasks()) > 2 { // connect task and setup-profiles tasks are 2 tasks, other tasks are hooks
-		// TODO: in preseed mode make interface hooks wait for mark-preseeded task.
-		for _, t := range autots.Tasks() {
-			if t.Kind() == "run-hook" {
-				return fmt.Errorf("interface hooks are not yet supported in preseed mode")
+	// If hooks are not present then connects can be executed during preseeding.
+	// Otherwise we will run all connects, their hooks and setup-profiles after
+	// preseeding (on first boot). Note, we may be facing multiple connections
+	// here where only some have hooks; however there is no point in running
+	// those without hooks before mark-preseeded, because only setup-profiles is
+	// performance-critical and it still needs to run after those with hooks.
+	if m.preseed && hasHooks {
+		for _, t := range st.Tasks() {
+			if t.Kind() == "mark-preseeded" {
+				markPreseeded := t
+				// consistency check
+				if markPreseeded.Status() != state.DoStatus {
+					return fmt.Errorf("internal error: unexpected state of mark-preseeded task: %s", markPreseeded.Status())
+				}
+
+				firstTaskAfterBoot, err := firstTaskAfterBootWhenPreseeding(snapsup.InstanceName(), markPreseeded)
```

Comment:
(for my own clarity) the reason this function works is because we specifically already checked if this snap has an install hook, so if we are in this branch at all, we know that this snap has an install hook

Reply:
mmm now I'm confused again, not sure that this function will always work or not

Reply:
I hope I clarified this in other comments.

Reply:
yes thank you